allennlp/models/encoder_decoders/nlvr/nlvr_coverage_semantic_parser.py

```diff
@@ -0,0 +1,375 @@
+                             label_strings=label_strings,
+                             agenda_data=agenda_data)
+        return outputs
+
+    def _get_checklist_info(self,
+                            agenda: torch.LongTensor,
+                            all_actions: List[ProductionRuleArray]) -> Tuple[torch.Tensor,
+                                                                             torch.Tensor,
+                                                                             torch.Tensor]:
+        """
+        Takes an agenda and a list of all actions and returns a target checklist against which the
+        checklist at each state will be compared to compute a loss, indices of ``terminal_actions``,
+        and a ``checklist_mask`` that indicates which of the terminal actions are relevant for
+        checklist loss computation. If ``self.penalize_non_agenda_actions`` is set to``True``,
+        ``checklist_mask`` will be all 1s (i.e., all terminal actions are relevant). If it is set to
+        ``False``, indices of all terminals that are not in the agenda will be masked.
+
+        Parameters
+        ----------
+        ``agenda`` : ``torch.LongTensor``
+            Agenda of one instance of size ``(agenda_size, 1)``.
+        ``all_actions`` : ``List[ProductionRuleArray]``
+            All actions for one instance.
+        """
+        terminal_indices = []
+        target_checklist_list = []
+        agenda_indices_set = set([int(x) for x in agenda.squeeze(0).data.cpu().numpy()])
+        for index, action in enumerate(all_actions):
+            # Each action is a ProductionRuleArray, a dict with keys "left" and "right", and
+            # values are tuples where the second element shows whether element is a
+            # non_terminal.
+            if not action["right"][1]:
+                terminal_indices.append([index])
+                if index in agenda_indices_set:
+                    target_checklist_list.append([1])
+                else:
+                    target_checklist_list.append([0])
+        # We want to return checklist target and terminal actions that are column vectors to make
+        # computing softmax over the difference between checklist and target easier.
+        # (num_terminals, 1)
+        terminal_actions = nn_util.new_variable_with_data(agenda,
+                                                          torch.Tensor(terminal_indices))
+        # (num_terminals, 1)
+        target_checklist = nn_util.new_variable_with_data(agenda,
+                                                          torch.Tensor(target_checklist_list))
+        if self._penalize_non_agenda_actions:
+            # All terminal actions are relevant
+            checklist_mask = torch.ones_like(target_checklist)
+        else:
+            checklist_mask = (target_checklist != 0).float()
+        return target_checklist, terminal_actions, checklist_mask
+
+    def _update_metrics(self,
```

Comment:
Maybe possible to share this code between the two classes, but probably not worth worrying about at this point.  Once we're to the cleaning-up-the-code stage, we can think about this more.

Reply:
Right. Added a TODO.